1단계 미션 제출합니다

owenoweny opened the pull request:
안녕하세요, 오웬 이요환입니다.

- 요구사항과 HTTP 표준을 고려했을 때 프로그램에서 커버해야하는 범위를 많이 고민하면서 작성했습니다.
- 이전 과제들에 비해서 요구사항이 크고 도메인이 복잡해서, OOP의 측면에서나 요구사항 만족 여부에서나 고민을 많이 했지만 완성도가 높지 않은 것 같습니다. 이 두 가지 측면에서 피드백 주시면 감사드리겠습니다.
- 테스트 코드를 아직 거의 작성하지 못했는데, 리팩토링하면서 추가하고자 합니다.

잘 부탁드립니다.

dongjoon1251 commented on src/main/java/webserver/http/HttpRequest.java:
```diff
@@ -60,15 +65,38 @@ public void setEntity(Map<String, String> entity) {
     }
 
     public Cookies getCookies() {
```
cookie를 처음에 파싱할 때 필드에 저장해두면 null체크 상관없이 바로 리턴해줄 수 있지 않을까요? :)

karian7 replied:
할때 Request Line 파싱 -> 헤더 파싱 -> 바디 파싱 의 절차로 진행되고 있어서 헤더 파싱하여 저장 하는 단계에 같이 설정하도록 바꿨습니다.

17faaeb 에서 수정하였습니다 😀

dongjoon1251 reviewed (changes requested):
테스트 코드도 자세히 작성해주시고 요구사항도 잘 정리해주셨네요 💯 
쿠키 설정 하는 부분 코멘트 드렸습니다. 
그리고 이전에 SEE_OHTER로 사용하셨던 것 같은데요. 저도 구현할때 적용하려고 보다보니 302는 Found로 표현되는 것 같습니다. 같이 확인 부탁드려요!

karian7 commented:
> 그리고 이전에 SEE_OHTER로 사용하셨던 것 같은데요. 저도 구현할때 적용하려고 보다보니 302는 Found로 표현되는 것 같습니다. 같이 확인 부탁드려요!

헉 그렇네요. 옆에 Status Value 를 `302 Found` 로 해놓고선.ㅎㅎ e2d49af7edfb99269af9adc042724613434bd059 에서 수정하였습니다.

dongjoon1251 commented:
마지막 단계까지 마무리 잘해주셨네요 ㅎㅎ
제가 미션을 먼저 수행해보지 못한 부분도 있고, 
피드백을 드리는 기준이 되는 설계가 없어서 좀더 다양한 피드백을 드리지 못해서 죄송합니다. 
고생하셨습니다!

owenoweny committed:
docs: 기능 목록 작성

owenoweny committed:
feat: HTTP 요청 메세지 파싱 기능 구현

owenoweny committed:
feat: 정적 페이지 GET 요청 응답 기능 구현

owenoweny committed:
refactor: URI 역할 분리

owenoweny committed:
feat: css 컨텐츠타입 처리 기능 구현

owenoweny committed:
refactor: Http 응답 클래스 역할 분리

owenoweny committed:
feat: 회원 저장 기능 구현

- ApiHandler: 어노테이션 기반으로 요청 라우팅 기능 구현
- HttpRequest.headers, body: value 자료형을 String으로 변경
-

owenoweny committed:
feat: /user/create 리다이렉트 기능 구현

owenoweny committed:
refactor:

- 변수 네이밍 수정
- 문자열 상수화
- 메소드 분리
- enum 패키지 분리, 예외 처리

owenoweny committed:
refactor: URI.extension 반환 방식 수정

- Optional을 반환하지 않도록 수정
- 존재하지 않은 경우 예외를 발생하도록 수정

owenoweny committed:
refactor: 문자열 상수화

owenoweny committed:
refactor: 가독성을 위한 메소드 분리 및 예외 메세지 수정

owenoweny committed:
fix: RequestHandler - 정적 리소스 요청 여부 검증 분기 버그 수정

owenoweny committed:
feat: 루트 페이지 api 구현

liquidjoo commented on src/main/java/enums/MIME.java:
```diff
@@ -0,0 +1,38 @@
+
+import exceptions.HttpRequestFormatException;
+
+public enum MIME {
```
enum 활용 👍

liquidjoo commented on src/main/java/enums/StatusCode.java:
```diff
@@ -0,0 +1,36 @@
+package enums;
+
+public enum StatusCode {
```
status, Code 관리 👍

liquidjoo commented on src/main/java/utils/FormUrlEncodedParsingStrategy.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class FormUrlEncodedParsingStrategy implements HttpBodyParsingStrategy {
```
인터페이스 활용 👍

liquidjoo commented on src/main/java/utils/HttpBodyParser.java:
```diff
@@ -0,0 +1,23 @@
+        if ("application/x-www-form-urlencoded".equals(contentType)) {
+            return new HttpBodyParser(new FormUrlEncodedParsingStrategy());
+        }
+        //예외를 얘가 던지는게 맞나?... HttpRequest가 던져야하지 않나?
```
주석은 PR시엔 제거 해주면 좋습니다~

liquidjoo commented on src/main/java/utils/HttpBodyParser.java:
```diff
@@ -0,0 +1,23 @@
+    public static HttpBodyParser from(String contentType) {
+        if ("application/x-www-form-urlencoded".equals(contentType)) {
+            return new HttpBodyParser(new FormUrlEncodedParsingStrategy());
+        }
```
정적 팩토리 메서드 패턴을 사용한다면 기본 생성자는 접근제어자를 통해 제어해보면 어떨까요?

liquidjoo commented on src/main/java/webserver/APIHandler.java:
```diff
@@ -0,0 +1,65 @@
+import java.util.Map;
+
+public class APIHandler {
+    private static Map<HttpMethod, URIHandlerMapping> requestMappings;
```
static 을 활용해 캐싱 처럼 활용하는 점 좋습니다 👍

liquidjoo commented on src/main/java/webserver/APIHandler.java:
```diff
@@ -0,0 +1,65 @@
+
+        for (Method method : methods) {
+            add(method);
+        }
```
초기화 영역이지만 반복문을 두 번 사용했다는 것은 여러 책임이 존재한다고 볼 수 있을 것 같은데요
메서드 분리를 통해 책임을 나누어 해보면 어떨까요?
어떤 장점이 있을까요?

owenoweny replied:
피드백 감사합니다~ 메소드 분리를 통해 가독성을 높일 수 있을 것이라고 생각해서 말씀해주신대로 리팩토링 진행해봤습니다.
  해당 부분 작성할 때는, **메소드를 분리함으로써 얻는 가독성에서의 이점**, **메소드를 분리함으로써 코드가 길어지고 메소드가 많아지는 문제** 사이에서 고민하다가 분리하지 않기로 결정했습니다. 로직이 어느정도 직관적이고 간단하다고 생각했기 때문입니다.

liquidjoo replied:
메소드를 분리한다는 뜻은 책임을 나누는 것인데요 책임이 많아지면 하나의 역할로 묶여 새로운 역할을 하는 객체가 나올 수 있을 것 같아요 :)
본인의 코드를 이해하기엔 직관적일 수 있으나 다른 사람이 코드를 본다면 어떻게 이해하면 좋을지 고민해 보면 좋을 것 같네요

liquidjoo commented on src/main/java/webserver/APIHandler.java:
```diff
@@ -0,0 +1,65 @@
+        for (HttpMethod httpMethod : HttpMethod.values()) {
+            requestMappings.put(httpMethod, new URIHandlerMapping());
+        }
+        Method[] methods = APIHandler.class.getDeclaredMethods();
```
Reflection 을 사용하셨군요
객체지향적 설계를 할 땐 좋은 방법은 아니지만 WAS 와 같은 어플리케이션을 만들 땐 어쩔수 없이 강요되는 부분이 있네요 😭 
하지만, 실무에서 사용할 땐 조심히 사용하는 것이 좋습니다~

liquidjoo commented on src/main/java/webserver/APIHandler.java:
```diff
@@ -0,0 +1,65 @@
+        DataBase.addUser(new User(userId, password, name, email));
+
+        return HttpResponse.found("http://localhost:8080/index.html");
+    }
```
해당 메서드는 마치 Controller 와 같은 역할을 하는 것 같네요~
하나의 역할을 만들어서 책임을 옮겨보면 어떨까요?

liquidjoo commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,89 @@
+                builder.body);
+    }
+
+    public static class Builder {
```
생성자가 있는데 빌더 패턴을 사용한 이유가 있을까요?
빌더 패턴을 사용한다면 기본 생성자는 접근제어자를 통해 제어해 보면 어떨까요?

owenoweny replied:
HTTP 요청은 파싱 로직이 복잡할 것이라고 생각해서 HttpRequest 객체 생성이 다양한 형태로 이루어질 수 있다고 생각했습니다. (객체 생성이 단계적으로 이루어지거나, 초기화가 필수적이지 않은 멤버가 있는 등) 다만 구현하고 나서는 빌더 패턴을 적용해서 얻을 수 있는 장점이 없어보이는 것 같습니다. 조언 감사합니다~

liquidjoo commented on src/main/java/webserver/HttpRequest.java:
```diff
@@ -0,0 +1,89 @@
+        private String protocol;
+        private HttpMethod httpMethod;
+        private Map<String, String> headers;
+        private Map<String, String> body;
```
로또에서 배웠던 일급컬렉션을 Header, Body 에도 적용해 볼 수 있을 것 같아요 :)

liquidjoo commented on src/main/java/webserver/HttpRequestParser.java:
```diff
@@ -0,0 +1,123 @@
+        String protocol = requestLine[2];
+
+        Map<String, String> headers = parseHeader(bufferedReader);
+        Map<String, String> body = parseBody(bufferedReader, headers);
```
일급컬렉션을 활용해 본다면 역할 생성, 책임 이동이 있을 것 같네요~

liquidjoo commented on src/main/java/webserver/HttpRequestParser.java:
```diff
@@ -0,0 +1,123 @@
+            stringURI = removeQuery(stringURI);
+            String extension = parseExtension(stringURI);
+            return new URI(path, parameters, MIME.from(extension.toUpperCase()));
+        }
```
다중 if 는 제거해 보면 좋을 것 같아요~

owenoweny replied:
조언 감사합니다~ if문이 별도로 있는 경우도 다중 if인가요?? 제거해야하는 이유는 가독성 때문일까요?

liquidjoo replied:
네 맞습니다 분기 처리가 계속 된다는 것은 객체지향적 설계로 보았을 때 어떤 단점이 생겨서 지양하고 있을까요?

liquidjoo commented on src/main/java/webserver/URI.java:
```diff
@@ -0,0 +1,50 @@
+
+public class URI {
+
+
```
```suggestion

```

liquidjoo commented on src/main/java/webserver/URI.java:
```diff
@@ -0,0 +1,50 @@
+
+
+    private final String path;
+    private final Map<String, String> parameters;
```
해당 부분도 일급컬렉션을 만들어 볼 수 있을 것 같아요~

liquidjoo reviewed (changes requested):
안녕하세요 요한님 :)
1단계 미션 잘해주셨어요 👍 
몇 가지 코멘트 남겼어요~ 확인하고 다시 요청 주세요 🙇

owenoweny committed:
feat: HttpRequestParser - URL 디코딩 로직 추가

owenoweny committed:
fix: list.html 템플릿을 받을 수 있도록 수정

owenoweny committed:
chore: 주석 삭제

owenoweny committed:
refactor: 가독성을 위한 메소드 분리

owenoweny committed:
refactor: 외부에서 사용하지 않을 생성자 접근제어자 수정

owenoweny committed:
refactor: API 요청 처리 클래스 역할 분리

owenoweny committed:
refactor: HttpRequest 빌더 클래스 삭제

owenoweny committed:
refactor: Http 파싱 로직 리팩토링

- 파싱 책임을 HttpRequestParser.java -> HttpRequest.java로 이동
- Http 헤더 정보를 나타내는 HttpHeaders 클래스 구현
- Http 요청 바디를 나타내는 HttpRequestBody 클래스 구현

owenoweny committed:
refactor: HTTP 메세지 관련 클래스 패키지 이동

owenoweny committed:
refactor: ResourceHandler.java -> StaticResourceHandler.java 리네임

owenoweny committed:
refactor: HttpRequest.parseHeader 메소드가 HttpHeaders를 반환하도록 수정

owenoweny commented:
@liquidjoo 피드백 주신 부분 반영하고 리팩토링했습니다~

- `HttpRequest`에서 `HttpHeaders`, `HttpRequestBody` 클래스를 분리하고, 관련된 파싱 로직을 응집성 있게 갖게 하고자 했습니다.
- 이에 따라 `HttpRequestParser` 클래스를 삭제했습니다.
- API 요청을 처리하는 (스프링의 Controller 구현체들의 역할을 하는) 메소드들을 따로 분리했습니다. (`RequestProcessor.java`)
- URL 디코딩 로직을 추가했습니다. 

리뷰 부탁드리겠습니다.
추가로, 코멘트에 남긴 질문에도 답변해주시면 감사드리겠습니다~

owenoweny committed:
refactor: RequestParameters - URI의 요청 파라미터 역할 분리

liquidjoo commented on src/main/java/webserver/httpmessage/HttpHeaders.java:
```diff
@@ -0,0 +1,69 @@
+
+import static webserver.httpmessage.HttpRequest.*;
+
+public class HttpHeaders {
```
피드백 반영 👍

liquidjoo commented on src/main/java/webserver/httpmessage/HttpRequest.java:
```diff
@@ -0,0 +1,84 @@
+    private final HttpHeaders headers;
+    private final HttpRequestBody body;
+
+    public HttpRequest(URI uri,
+                       HttpMethod httpMethod, String protocol,
+                       HttpHeaders headers,
+                       HttpRequestBody body) {
+        this.uri = uri;
+        this.httpMethod = httpMethod;
+        this.protocol = protocol;
+        this.headers = headers;
+        this.body = body;
+    }
```
정적팩토리 메서드 패턴을 사용한다면 생성자 접근제어자를 활용해 제어해 보아요 :)

liquidjoo commented on src/main/java/webserver/httpmessage/HttpRequest.java:
```diff
@@ -0,0 +1,84 @@
+        return new HttpRequest(uri, httpMethod, protocol, headers, httpRequestBody);
+    }
+
+    private static HttpRequestBody parseBody(BufferedReader bufferedReader, HttpHeaders headers) throws IOException {
+        HttpRequestBody httpRequestBody = HttpRequestBody.empty();
+        if (headers.containsKey(CONTENT_TYPE_KEY) && headers.containsKey(CONTENT_LENGTH_KEY)) {
+            int contentLength = Integer.parseInt(headers.get(CONTENT_LENGTH_KEY));
+            String bodyString = IOUtils.readData(bufferedReader, contentLength);
+            httpRequestBody = HttpRequestBody.of(headers.get(CONTENT_TYPE_KEY), bodyString);
+        }
+        return httpRequestBody;
+    }
```
`HttpRequest` 에 어울리는 책임일까요?

liquidjoo commented on src/main/java/webserver/httpmessage/HttpRequestBody.java:
```diff
@@ -0,0 +1,26 @@
+
+import java.util.Map;
+
+public class HttpRequestBody {
+    private final Map<String, String> values;
```
의미있는 일급 컬렉션 활용 👍

liquidjoo commented on src/main/java/webserver/httpmessage/RequestParameters.java:
```diff
@@ -0,0 +1,29 @@
+public class RequestParameters {
+    private final Map<String, String> values;
+
+    public static RequestParameters from(String stringUri) {
+        int queryStartIndex = stringUri.indexOf(URI.QUERY_SEPARATOR) + 1;
+        if (queryStartIndex == 0) {
+            return new RequestParameters(Map.of());
+        }
+        String query = stringUri.substring(queryStartIndex);
+        Map<String, String> parameterMap = Arrays.stream(query.split(HttpRequest.PARAMETER_SEPARATOR))
+                .map(str -> str.split(URI.PARAMETER_EQUAL_SIGN))
+                .map(strings -> Map.entry(strings[0], strings[1]))
+                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        return new RequestParameters(parameterMap);
+    }
```
내부 상태 값인 values 를 사용하고 있지 않은 것 같은데요 
객체의 응집도가 많이 떨어지는 것 같네요
어떻게 수정하면 될까요?

liquidjoo commented on src/main/java/webserver/httpmessage/URI.java:
```diff
@@ -0,0 +1,90 @@
+    private final RequestParameters parameters;
+    private final Optional<MIME> extension;
+
+    public URI(String path, RequestParameters parameters) {
+        validatePathFormat(path);
+        this.path = path;
+        this.parameters = parameters;
+        this.extension = Optional.empty();
+    }
+
+    public URI(String path, RequestParameters parameters, MIME extension) {
+        this.path = path;
+        this.parameters = parameters;
+        this.extension = Optional.of(extension);
+    }
+
+    public static URI from(String stringURI) {
+        RequestParameters parameters = RequestParameters.from(stringURI);
+        String path = URLDecoder.decode(stringURI, StandardCharsets.UTF_8);
+
+        if (hasQuery(stringURI)) {
+            int queryStartIndex = stringURI.indexOf(QUERY_SEPARATOR);
+            path = stringURI.substring(0, queryStartIndex);
+        }
+        if (hasExtension(path)) {
+            stringURI = removeQuery(stringURI);
+            String extension = parseExtension(stringURI);
+            return new URI(path, parameters, MIME.from(extension.toUpperCase()));
+        }
+        return new URI(path, parameters);
+    }
```
주생성자, 부생성자, 정적팩토리 메서드 패턴을 혼합하여 사용할 경우 어떤 일들이 발생할 수 있을까요?

liquidjoo reviewed (approved):
안녕하세요 요환님 :)
1단계 피드백 반영 잘해주셨어요 👍 
몇 가지 코멘트 남겼는데요~ 
확인하고 다음 단계에 반영해 보면 좋을 것 같아요
다음 단계 진행해 주세요 🙇
